### ISSUE TITLE
How exactly Notification Delegation does work?

### ISSUE THREAD

**Issue by FluorescentHallucinogen** [GoogleChromeLabs/svgomg-twa#60]
From the [description](https://github.com/GoogleChrome/custom-tabs-client/blob/26aa6744970790de086e92014aebf543b4585fb7/customtabs/src/android/support/customtabs/trusted/TrustedWebActivityService.java#L56-L94), I don’t quite understand how Notification Delegation works.

Could you please explain?

If I set `enableNotifications` to `true`, does this mean that notifications will be managed by TWA wrapper and Chrome notification permission prompts will not be displayed (because TWA wrapper is Android native app and notifications in Android native apps are enabled by default without permission prompts)?

If yes, how does this fit with the [recent notification experiment in Chrome](https://twitter.com/alexey_rodionov/status/1191080942011596804)?

**Comment by PEConn** [comment]
You've got it correct.

The first time the TWA is run, we enable Chrome notifications for the associated origin. If the TWA has notifications disabled in Android Settings, we disable them in Chrome instead. There's a little bit of latency with how this gets propagated, but things should get updated on the next TWA launch at latest.

When the TWA is uninstalled, the notification permission is set to whatever it was before it was installed.

I don't know about those experiments 100%, but this should take precedence over them. (It looks like the experiments are to do with how the notification prompt is displayed to the user - we just set the notification permission, not requiring a user prompt at all.)

**Comment by FluorescentHallucinogen** [comment]
@PEConn 

Thanks for clarification!

By the way, what about delegation of other permissions (like geolocation, etc.), not only notifications? Is it possible?

PTAL at https://crbug.com/974785 for motivation.

TL;DR: It's very difficult, if not impossible, to explain to ordinary users (not developers) that in order to change permissions for the _app_ they need to go to the _browser_. This is not obvious and inconvenient (hidden deep in the browser settings). Most ordinary users don't distinguish PWA from Android native apps (they look and feel like native Android app: icon on the home screent, standalone mode, etc.). It would be really nice to integrate PWA permissions more deeply into OS (into Android app settings). This is the first place users try to open to change permissions.

**Comment by PEConn** [comment]
Yeah, we're planning on expanding permission delegation to include permissions other than notifications.

Since each different permission needs to be plumbed through manually, we're going to be starting with the bigger ones (Geolocation and maybe Camera, although that's a complex API) and then evaluating the payoff for the rest. The other difficulty is that while notifications are enabled by default, the rest aren't, so when a website requests a permission, we need to pipe that through to the client app and have that display the Android UI.

**Comment by FluorescentHallucinogen** [comment]
@PEConn I've partially fixed the problem described in https://github.com/GoogleChromeLabs/svgomg-twa/issues/60#issuecomment-549781939 by #61.

**Comment by FluorescentHallucinogen** [comment]
@PEConn 
>Yeah, we're planning on expanding permission delegation to include permissions other than notifications.

Any progress on that?

**Comment by PEConn** [comment]
We're going to be looking into Geolocation this quarter (no guarantees on when it will be finished).

Is there any specific API you're waiting on?

**Comment by FluorescentHallucinogen** [comment]
> Is there any specific API you're waiting on?

I look forward for geolocation permission delegation.

The only nuance that I see is that we should cover the case when geolocation disabled in the OS/device settings (not Android app settings):

![1](https://user-images.githubusercontent.com/7892779/68206457-c9341600-ffdd-11e9-854c-49ca0b3c0955.png)
![2](https://user-images.githubusercontent.com/7892779/68206462-cb967000-ffdd-11e9-8938-85f975a43d34.png)
![3](https://user-images.githubusercontent.com/7892779/68206466-cdf8ca00-ffdd-11e9-8e43-512fa48aa280.png)
![4](https://user-images.githubusercontent.com/7892779/68206471-d05b2400-ffdd-11e9-9cbc-02e39dbac484.png)

**Comment by FluorescentHallucinogen** [comment]
Also very interested how Geolocation Delegation will fit with background geolocation (https://joebirch.co/2019/03/18/exploring-android-q-location-permissions/)?

**Comment by PEConn** [comment]
We don't have any answers for you now, but should do in the coming months.

**Comment by tobiasharle** [comment]
> The first time the TWA is run, we enable Chrome notifications for the associated origin. If the TWA has notifications disabled in Android Settings, we disable them in Chrome instead. There's a little bit of latency with how this gets propagated, but things should get updated on the next TWA launch at latest.

Are you saying that Notification.permission should return true without the need to call Notification.requestPermission?

Because that is not the case for me. I still have to request the permission even though I have my TWA setup with a DelegationService working. It is receiving the notifications I'm sending.

**Comment by PEConn** [comment]
Yes, that should be the case.

Are you setting `enableNotifications` to true in the build.gradle? And
verification is completing successfully?

Other than that, are you requesting the permission the instant the user
loads the page? There may be a delay with the propagation, so maybe wait
for a user interaction.

**Comment by FluorescentHallucinogen** [comment]
@PEConn @andreban I'm concerned that Notification Delegation could be used to circumvent [Quieter Notifications](https://blog.chromium.org/2020/01/introducing-quieter-permission-ui-for.html).

**Comment by PEConn** [comment]
Your link doesn't talk about users getting too many notifications, but
about users getting too many notification permission prompts. Notification
Delegation helps with this because an installed TWA doesn't have to ask for
the notification permission, it gets it automatically.

Yes, it's true that a user installing a TWA can now get push notifications,
but this aligns with how native Android apps work and we're trying to
reduce the differences between TWAs and native apps. Also, app installation
is seen as a stronger signal than accepting the notification permission.

**Comment by andreban** [comment]
Quieter Notifications aim to fix the issue where sites ask for notification permission even before a user engages with it for the first time (eg: on the first visit).

TWAs have the same behaviour as native Apps and get the notification permission by default. I think the concern here would be for developers to abuse and spam users with notifications. But this frustrates those who could be the most engaged users, and leads to the user uninstalling the app.

**Comment by FluorescentHallucinogen** [comment]
> TWA doesn't have to ask for
the notification permission, it gets it automatically.

I mean that some sites can abuse (misuse) this fact.

I believe the problem is not as simple as it seems at first glance.

> installation is seen as a stronger signal than accepting the notification permission.

> Quieter Notifications aim to fix the issue where sites ask for notification permission even before a user engages with it for the first time (eg: on the first visit).

Installation is even before the first visit. So using installation as a strong signal for accepting permission is bad idea. If the user installed the app (wrapper), it doesn't mean that he even started it, saw the content and ingaged with it. And it doesn't mean that he wants to grant some permissions (e.g. receive notifications by default) at all. The process of disabling notifications may be unintuitive (non-obvious) for some users.

> this aligns with how native Android apps work and we're trying to reduce the differences between TWAs and native apps.

Ok, but what about delegation of other permissions like WebNFC that deal with user data, user identity or other privacy sensitive attributes (read more: https://w3c.github.io/web-nfc/#security)?

Note, that on Android there is no permission request for NFC (it's enabled by default).

**Comment by andreban** [comment]
Currently, the **only** permission granted by default on TWAs is the notification permission. Other works like in the Web. We are looking into making other permissions consistent to native apps, but analyzing case-by-case.

We are definitely going to keep track on how the this is being used by developers.  Android itself is also constantly looking into how native apps are using this, and working on giving users more control.

@b1tr0t FYI

**Comment by tobiasharle** [comment]
> Are you setting `enableNotifications` to true in the build.gradle?
No, but I have added a service (overriding the DelegationService) to my manifest and it's always enabled and exported. It successfully receives web push notifications. Is there something more to the enableNotifications flag that I'm not seeing?

> And verification is completing successfully?
What verification is that?

> Other than that, are you requesting the permission the instant the user
> loads the page? There may be a delay with the propagation, so maybe wait
> for a user interaction.
It's done on user interaction

**Comment by PEConn** [comment]
@tobiasharle That is strange - if your TWA is getting notifications through notification delegation, then auto-enrolment should also be working (unless you're using a version of Chrome before 75).

Is the notification permission for the TWA in your Android settings not blocked? Are you doing anything fun in your override of DelegationService?

Either way if I could see your code or an example APK, that would be helpful.

**Comment by tobiasharle** [comment]
@PEConn For the moment, I reverted to not overriding DelegationService to avoid any potential sources of error.
It seems as there is some kind of connection between the site and the app but it is not working as I expect it to so let me break it down in detail:

1. I install my TWA app and it navigates to my site.
2. The site checks the value of `Notification.permission`. It is _"default"_ (**Expected value: _"granted"_** since the app is installed)
3. `Notification.requestPermission` is called and shows the permission dialog. It is accepted and push notifications are setup.
4. I check the site settings in chrome. It says _Notifications - Allow_
5. A push notification is sent and received successfully in the app through DelegationService.
6. I disable the app notification channel called _General_ that was automatically created.
7. Another push notification is sent and as expected, it is not received.
8. I check the site settings in chrome again and now it says _Notifications - Managed by "Name of my app"_ :tada:
9. I enable the _General_ app notification channel again.
10. Another push notification is sent and it's not received. (**Expected: It should be received**)
11. I debug the site and check the value of `Notification.permission`. It is _"denied"_ (**Expected value: _"granted"_**)

**Comment by tobiasharle** [comment]
@PEConn I have found the problem! In my intent-filter I was specifying a pathPattern like so:
```
<intent-filter
    android:autoVerify="true">
    <data
        android:scheme="https"
        android:host="Web site host"
        android:pathPattern="/.*" />
    <action android:name="android.intent.action.VIEW" />
    <category android:name="android.intent.category.BROWSABLE" />
    <category android:name="android.intent.category.DEFAULT" />
</intent-filter>
```
Once I removed the pathPattern attribute everything worked smoothly!

Is that to be expected or is it a bug?

**Comment by PEConn** [comment]
That is strange - it's probably a bug, but I need to have a think about what `pathPattern` does and whether it makes sense.

Why were you using it?

**Comment by PEConn** [comment]
(tracking the Chrome side of this bug at https://crbug.com/1050514 )

**Comment by tobiasharle** [comment]
I believe it was added to make sure that all clicked links with any sub-path was handled by the app and not opened in the browser.

**Comment by andreban** [comment]
Closing the issue here, as the question has been answered and the issue is being tracked on Chromium Bugs

**Comment by FluorescentHallucinogen** [comment]
@PEConn @andreban @b1tr0t Will permission delegation (for notifications and/or geolocation) work for WebAPK (not only TWA)?

**Comment by PEConn** [comment]
It doesn't at the moment - implementing it is do-able, but not on our current roadmap. I'll leave it to @b1tr0t to determine prioritization. 

The notification permission for TWAs is enabled by default because that is the norm for Android apps. For WebAPKs though, since they're installed through Chrome, not through the Play Store, maybe we need to consider different user expectations.

**Comment by FluorescentHallucinogen** [comment]
@PEConn Thanks for your reply!

Why am I asking? I see a lot of issues with "[WebAPK/TWA merging]" prefix in title in https://crbug.com. My question is not about _notification_ permission delegation, but about permission delegation in general (e.g. for upcoming geolocation permission delegation). The fact that in order to change permissions for the _app_ users need to go to the _browser_ is bad UX and affects both TWAs and WebAPKs.

In connection with the upcoming permissions updates in Android 11, this question is becoming more relevant. (See https://developer.android.com/preview/privacy/permissions for more details.)

E.g. I expect that enabling permission delegation for WebAPKs will enable one-time permissions and auto-reset permissions from unused apps features from upcoming Android 11 automatically. 👌

Half a year ago I asked about background geolocation (see https://github.com/GoogleChromeLabs/svgomg-twa/issues/60#issuecomment-574131917).

Your answer was:

> We don't have any answers for you now, but should do in the coming months.

Any news in the meantime?

I see that permission delegation for geolocation is almost done (see https://chromium-review.googlesource.com/q/delegation+owner:eirage%2540chromium.org). Right? Any ETA?

In connection with the upcoming location updates in Android 11, this question is becoming more relevant. (See https://developer.android.com/preview/privacy/location and https://developer.android.com/training/location/background for more details.)

I believe with the upcoming location updates in Android 11, implementing background geolocation support for Web will be safe. OS take care of things.

**Comment by PEConn** [comment]
Yeah, so we went to a fair amount of effort over the last year to combine the Chrome code backing TWAs and WebAPKs and while that would definitely make extending permission delegation to WebAPKs easier, launching that would be a product question, so we're not going to turn it on without some thought.

We've been working on location delegation - you can see some of the progress [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1069506) - with an optimistic target of M85 (though that's not set in stone). We also still need to finalize the TWA side code (over in android-browser-helper).

**Comment by TahaSaifee53** [comment]
Hello,

I had built apk from TWA with notification delegation on. Now, I have integrated fcm to send the notification, and the android apk is getting the notification from web push into apk also. But the tap on notification could not open the apk.(The notification is not from chrome but now it's coming from apk) 

Please suggest any way to open apk directly on click of web push from apk.
